### ISSUE TITLE
The local branch master is behind the remote one, therefore a new version won't be published

### ISSUE THREAD

**Issue by thopaw** [semantic-release/semantic-release#1208]
## Current behavior
I have a pipeline which is running for some time.
It runss unit tests, it deploys the current commit into a test stage, it runs e2e against it and then waits for a manual approval.

When it is approved, semanic-release is started to build tag the version which should go into production.

When in the meantime there are new commits on the master branch, the manual approved version can not be deployed anymore as semantic release is throwing the following error:

> The local branch master is behind the remote one, therefore a new version won't be published

This comes from index.js: 68
```
if (!(await isBranchUpToDate(options.branch, {cwd, env}))) {
        logger.log(
          `The local branch ${options.branch} is behind the remote one, therefore a new version won't be published.`
        );
        return false;
      }

```

An every build which is not on the latest version of a branch semantic-release exits with the error explained above.

## Expected behavior

In longer running pipelines or when a certain version is approved manually the execution of a deicated commit should not run in a failure but build this dedicated version.

## Environment

- **semantic-release** version:
    ```
"@semantic-release/changelog": "^3.0.4",
    "@semantic-release/gitlab": "^3.1.5",
    "semantic-release": "^15.13.16",
    "semantic-release-slack-bot": "^1.0.4",
```
- CI environment: gitlab
- Plugins used: gitlab
- **semantic-release** configuration: 

```
branch: master
plugins:
  - '@semantic-release/commit-analyzer'
  - '@semantic-release/release-notes-generator'
  - '@semantic-release/gitlab'
  - - semantic-release-slack-bot
    - notifyOnSuccess: true
      notifyOnFail: true

```
- CI logs:
```
[14:04:55] [semantic-release] › ℹ  Running semantic-release version 15.13.16
[14:04:55] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/gitlab"
[14:04:55] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[14:04:55] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
[14:04:55] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/gitlab"
[14:04:55] [semantic-release] › ✔  Run automated release from branch master
[14:04:56] [semantic-release] › ℹ  The local branch master is behind the remote one, therefore a new version won't be published.

```

**Comment by pvdlg** [comment]
This is the intended behavior. There is many reasons for that but the main one is that some plugins can affect the repository on release, for example `@sematnic-release/git` create a new commit with the updated `package.json` and `CHANGELOG.md`.
If the build is not running on the latest commit this would fail.

Even is semantic-release exit with this message, the following run will trigger the release.
So instead of releasing 1.0.1 with commit X and then 1.0.2 with commit Y a few minutes later, you will just release 1.0.1 with both commits. How does that make a difference in your case?

**Comment by thopaw** [comment]
> There is many reasons for that but the main one is that some plugins can affect the repository on release, for example @sematnic-release/git create a new commit with the updated package.json and CHANGELOG.md.

This is defintly a point.

> So instead of releasing 1.0.1 with commit X and then 1.0.2 with commit Y a few minutes later, you will just release 1.0.1 with both commits. How does that make a difference in your case?

The problem I see is that the pipeline is running for a certain time (~8 minutues) and then waits for a manual approval of the business. When in the time of the start of the pipeline and the approval of the version a new commit is pushed the pipeline will fail.

I guess I can work around it by tagging on the approval and start a new pipeline on that tag and let run semantic release on that tag. For branch I can also use a tag when I do push during the release, right?

**Comment by pvdlg** [comment]
> The problem I see is that the pipeline is running for a certain time (~8 minutues) and then waits for a manual approval of the business. When in the time of the start of the pipeline and the approval of the version a new commit is pushed the pipeline will fail.

The build shouldn't fail. In this scenario semantic-release will release with this message but will return `0` which a success return code.

> I guess I can work around it by tagging on the approval and start a new pipeline on that tag and let run semantic release on that tag. For branch I can also use a tag when I do push during the release, right?

I don't really understand what you mean by that. That seems quite complicated.
Just wait for the next build to run and publish. What's the difference if instead of 2 publishes within 8 minutes you get one that contains the commit of both?

**Comment by thopaw** [comment]
> The build shouldn't fail. In this scenario semantic-release will release with this message but will return 0 which a success return code.

Yes, but nothing gets released

> What's the difference if instead of 2 publishes within 8 minutes you get one that contains the commit of both?

Yes, when the second pipeline runs through before another commit is pushed.
I guess the easiest way is to introduce a development branch or to work with tags for build.

**Comment by r4j4h** [comment]
I know this is closed, but I wanted to clarify for @pvdig:

> What's the difference if instead of 2 publishes within 8 minutes you get one that contains the commit of both?

By merging the changes it can complicate versioning for downstream users as well as make reverting things unnecessarily more complex at the commit level rather than tag level. More importantly, we just encountered a situation where a fix was running, and someone from another team released a feature.

I was robbed of my tag including only my fix, and now only have a version with both things combined. As you imply, this is a great feature. But for a downstream user who only wanted the fix and not the feature they are out of luck.

**Comment by r4j4h** [comment]
Also, this was supposed to have been fixed in https://github.com/semantic-release/semantic-release/issues/993 but based on https://github.com/semantic-release/semantic-release/issues/709 still seems to be an issue.

**Comment by pvdlg** [comment]
So what would you propose to solve that issue?

**Comment by r4j4h** [comment]
I think it'd be the same fix as in #993 but on whatever code executes the `--dry-run` scenario.

From a glance I am guessing https://github.com/semantic-release/semantic-release/blob/7a939a897097a730f1579002ed641a583d3381c4/lib/git.js#L121 even though its name seems to imply it is the wrong function, but I don't see any other push functions or areas where `--dry-run` is being added.

**Comment by pvdlg** [comment]
That would just remove the code that make the verification.
You would still end up with an error when pushing the tag in case the local branch branch is behind.
See https://github.com/semantic-release/semantic-release/issues/1208#issuecomment-503199090

You can't push a commit if the local branch is behind the remote one.

**Comment by pvdlg** [comment]
Also I don't understand why you are referring to the `--dry-run` option. That seems completely unrelated to the problem you explained in https://github.com/semantic-release/semantic-release/issues/1208#issuecomment-555151554

**Comment by r4j4h** [comment]
Hmmm, I am not sure of the resolution after all. My point was around https://github.com/semantic-release/semantic-release/issues/993#issuecomment-440741417, but after some thinking I see your point that semantic-release is trying to push the change in package.json to set the version to the fix's semver fails because master has new changes. To that I see two ways around:

1. Push the tag with the fix semver, do not have master have the fix semver present. I see why someone might squirm at this, but I think npm would be fine with it. Checking out to the tag or the specific commit should still work fine as well.

2. Merge and push it. While not trivial, it's not impossible to determine if the only conflict is package.json's version line or if there are no conflicts there, and then several different git actions could be taken, potentially including a series of commits. This approach has been discussed and has other problems, such as how the pipeline that happened to run in parallel also has semantic-release that will run and how they interact must be made to work well also. It would not be clean, but I can hazily imagine a series of commits that does some reverts, makes the change it wants, unreverts sans package change, then the subsequent semantic-release does the same - it would involve a ridiculous amount of commits just for the swap-arounds but theoretically would work and shelling out to `git` makes most of those steps not as difficult as they would be if we had to do them manually.

I referred to the dry run because with my specific error it is where the problem occurs. I will post it at the end of this post. I was hoping it was as simple as a fix as https://github.com/semantic-release/semantic-release/issues/993#issuecomment-440741417 was but I see it's a little different.

Here's my specific error, although it's not so useful now that I've realized the package.json semver push problem. I guess I need to chew on that behavior for a while, I remember someone asking for a `--force` option and maybe that's what is needed.. just with a better name like `--only-version-tag` or `--ignore-master-branch`.

```
[6:16:21 PM] [semantic-release] › ✔  Run automated release from branch master
2019-11-18T18:16:22.231Z semantic-release:git Error: Command failed with exit code 1: git push --dry-run git@gitlab.webpt.com:emr-documentation/patient-chart-ui.git HEAD:master
    at makeError (/builds/emr-documentation/patient-chart-ui/node_modules/execa/lib/error.js:56:11)
    at handlePromise (/builds/emr-documentation/patient-chart-ui/node_modules/execa/index.js:114:26)
    at process._tickCallback (internal/process/next_tick.js:68:7)
2019-11-18T18:16:22.911Z semantic-release:git Error: Command failed with exit code 1: git push --dry-run https://gitlab-ci-token:[secure]@gitlab.webpt.com/emr-documentation/patient-chart-ui.git HEAD:master
    at makeError (/builds/emr-documentation/patient-chart-ui/node_modules/execa/lib/error.js:56:11)
    at handlePromise (/builds/emr-documentation/patient-chart-ui/node_modules/execa/index.js:114:26)
    at process._tickCallback (internal/process/next_tick.js:68:7)
[6:16:23 PM] [semantic-release] › ℹ  The local branch master is behind the remote one, therefore a new version won't be published.
```

**Comment by r4j4h** [comment]
I won't edit it so that things still make sense, but I didn't actually configure semantic-release on our projects so I started reading about the config. I just learned it is sensitive to the release branch's most recent commit being in tags. I think that invalidates my 2nd thought above.

**Comment by pvdlg** [comment]
Oh ok I though you were referring to the semantic-release `--dry-run` option.
Indeed we use the Git `--dry-run` option verify the current state of the local repo and exit with a proper message, rather than failing during the push step.

> Push the tag with the fix semver, do not have master have the fix semver present. I see why someone might squirm at this, but I think npm would be fine with it. Checking out to the tag or the specific commit should still work fine as well.

What the `@seamtnic-release/git` plugin does is not only about pushing the updated `package.json`. It's also used to push build assets and a `CHANGELOG.md`.
Pushing the updated `package.json`, `CHANGELOG.md` and build assets was by far one of the most requested feature we had. So preserving the ability to do that seems more important than the case you described in https://github.com/semantic-release/semantic-release/issues/1208#issuecomment-555151554. It is a legitimate case but seems far less common.

> Merge and push it. 

That doesn't seem to be realistically possible for several reasons:
- Not sure it would be possible to re-arrange the commits properly in every situation
- Would require a force push (in order to insert the release commit in between commits on the remote) and I don't think it's reasonable to ask users to allow force push on their master branch
- Not sure it would be possible with CIs that creates a clone with a detached head

**Comment by r4j4h** [comment]
Thanks for the response! I see your confusion on the `--dry-run`. Sorry about that, it was my fault due to my ignorance with semantic-release itself. I didn't know it had one so I didn't know to make the distinction clear. 😊 

> So preserving the ability to do that seems more important than the case you described in #1208 (comment). It is a legitimate case but seems far less common.

That makes a lot of sense.  The case I am describing is definitely a rarer edge case. We can still revert things by commit, and if we group commits in a branch before the merge it equates to the same effort. For downstream users I think we will just wash our hands until someone asks, and if they do then we can manually cut a fix. I doubt it will even be necessary.

I am not sure of a good way to actually prevent the issue though, we have several teams pushing into the same repository and no gatekeeper to wait. We could possibly teach devs to wait if there is an ongoing pipeline, and since we are using GitLab there may be a way to restrain the number of concurrent pipelines that run. Ultimately I agree and I think letting it just happen is fine. In the rare edge case there are plenty of manual solutions to fix it. :+1:

> That doesn't seem to be realistically possible for several reasons: [snip]

I agree. There might be some ways without requiring force push, but I think they would interfere with CIs with detached heads unless we took lots of special care. I don't think it'd be worth it, especially with how gross it would look in people's commit histories - I think they'd hate it hehe

**Comment by pvdlg** [comment]
> (EDIT) I will edit this one so as to not spam. I also just saw in the FAQ that the behavior of it pushing the package.json change is not default. That kind of makes both my 1st thought above as well as this entire problem invalid because it means my config is the cause of the problem because it is pushing up the changed package.json.

> I guess I want it to conditionally do that, or to not fail if there is a problem of it being behind. I don't see any ability to do that sort of thing on https://github.com/semantic-release/git. Maybe this would be better served as a feature request to that repo?(EDIT) I will edit this one so as to not spam. I also just saw in the FAQ that the behavior of it pushing the package.json change is not default. That kind of makes both my 1st thought above as well as this entire problem invalid because it means my config is the cause of the problem because it is pushing up the changed package.json.

So making the commit is done by `@semantic-release/git` indeed and is optional because it's a plugin you can choose to use or not.

Possibly we could figure out a way where we verify if we are behind the remote only in `@semantic-release/git`  and push the commit there. That would allow us to be able to push only the tags in semantic-release core.

I see one problem with that though. We want to make sure in semantic-release core that we have the permission to push tags (as in make sure the credential allows us to do that). This is why we use `git push --dry-run`. However this also fails if we are behind the remote.
We would need to find a command that allow to only verify the push permission (without failing if we are behind remote). Maybe `git push --dry-run --tags` would work. I need to test that.

**Comment by pvdlg** [comment]
So `git push --dry-run --tags` works. The problem is that if we move the verification to `@semantic-release/git` there is no way to stop the semantic-release process without erroring out.

Basically plugins can either succeed (resolve Promise) or fails (reject Promise). In the case where we are behind the remote branch we want to exit without errors.
That would require to create a mechanism for plugins to communicate that to the core.

In addition we would to call the `verifyCondition` passing the authenticated `repositoryUrl` which we do only for following plugins calls at the moment.

**Comment by pvdlg** [comment]
Another problem if we allow releases from outdated local branch is a possible race condition.
If the last release on the branch is `1.0.0` and we merge to `fix` commits quickly then the 2 CI jobs will race each other and might both try to release `1.0.1` almost at the same time leaving you in a very difficult to solve situation.

With the current code only one release can happens at the same time, so it avoid the race condition issue.

**Comment by r4j4h** [comment]
Thanks for the follow ups! That is great that I wasn't too far off base from what is possible. It seems like we are on the right trail for a solution.

I think having to increase the API surface of plugins isn't ideal, but having richer core-plugin communication could be beneficial and possibly enable plugin-plugin communication.

> Another problem if we allow releases from outdated local branch is a possible race condition.
If the last release on the branch is 1.0.0 and we merge to fix commits quickly then the 2 CI jobs will race each other and might both try to release 1.0.1 almost at the same time leaving you in a very difficult to solve situation.

Oh man this is such a good point. 🤦‍♀ You are absolutely right, if in my case the other team had also done only a fix-level change, what versions would I expect?

Thinking it through, I think semantic-release we would have to detect the tags and bump versions more and try again until it succeeds.

For example: if we were at 1.0.0, my fix (assuming it was first) would become 1.0.1, and then their fix would race and try to do 1.0.1 but would detect 1.0.1 exists and then try 1.0.2.

(Versus the current behavior of my fix bailing after it encounters the newer commit. Then their fix includes mine and theirs in 1.0.1.)

I think it would still leave a failure case open, or at least a weird behavior, where different (more likely if more than 2) versions could race and potentially re-order themselves depending on the outcome of their racing. E.g. 1.0.0, my fix could become 1.0.1, but while their fix is determining to use 1.0.2, ANOTHER third fix could be determining to use 1.0.2 as well. One will likely reach git remote first, and In that case the remaining one could fail again and try to increment to 1.0.3 and try again and would succeed.

That might be acceptable, but definitely sounds gross. Like you say, the current sequential way avoids the whole mess altogether.

**Comment by pvdlg** [comment]
> Oh man this is such a good point. 🤦‍♀ You are absolutely right, if in my case the other team had also done only a fix-level change, what versions would I expect?

That's unpredictable. It would depends on the specific in which the repo is cloned and tags pushed.

> Thinking it through, I think semantic-release would have to detect the tags and bump versions more and try again until it succeeds.

I don't think so. There can be multiple reason for a tag conflict. Just assuming it's due to a race condition and increasing the version is dangerous. In addition it can create the weird situation were a 1.0.1 and 1.0.2 can be created out of order from the git history perspective which would probably create other issues.


Overall I think it's probably better to leave the code as is and to find external solution such as limiting the job that make the release to 1 concurrent instance if possible.

**Comment by priyajeet** [comment]
We currently have this issue. We only have 1 build that runs, nothing parallel. But our builds take a while, so by the time semantic release tries to run, it fails because someone else pushed to the repo. Wondering if a config or option could be added to ignore this.

**Comment by nodkz** [comment]
@priyajeet I had the same problem. So to debug this I ran locally `./node_modules/.bin/semantic-release --dry-run`  and got an error that I trying to load status from wrong git-repository (hopefully I don't have access rights):
```
{ Error: Command failed with exit code 128: git ls-remote --heads https://***.kz/packages/app-server.git
remote: HTTP Basic: Access denied
```

So I found that wrong repo was written in my `package.json`:
```diff
"repository": {
    "type": "git",
-    "url": "https://***.kz/packages/app-root.git"
+    "url": "https://***.kz/packages/app-server.git"
  },
```

**Comment by rafalmaciejewski** [comment]
Is there any way to make semantic-release exit with code other than 0 when i get this message?

**Comment by agaudreault-jive** [comment]
We also hit that issue today as more of our teams are adopting semantic-release and continuous delivery. 

Basically, we have a Jenkins pipeline that runs with `disableConcurrentBuilds()` on each commit to master branch to make sure we do not run two builds at the same time. At the beginning of the pipeline, we use semantic-release in dry-run to get the next release version and use this version to build and publish our artifacts. Then, we run semantic-release again without dry-run to create the tag. 

Unfortunately, the run failed with `The local branch master is behind the remote one, therefore a new version won't be published.` because another pull request was merged during the build.

Even if the next build would create a release with the two commits, our build fails because we end up trying to override our artifacts pushed with the same version by the previous build.

We are also not using the @semantic-release/git plugin because our master branch is protected and pull request are enforced.

So I think it is a valid use case to have an option we can configure to allow semantic-release to push tags from a local branch behind the remote that would be disabled by default to keep the backward compatibility.

**Comment by agaudreault-jive** [comment]
@pvdlg Can you reopen this issue or should I create another one ?

**Comment by pvdlg** [comment]
> @pvdlg Can you reopen this issue or should I create another one ?

I think answers were provided and it was explained why what you request is not possible, but I can further comment.

> At the beginning of the pipeline, we use semantic-release in dry-run to get the next release version and use this version to build and publish our artifacts. Then, we run semantic-release again without dry-run to create the tag.

I don't know why you do that, maybe working around some Jenkins limitation, but we won't make any change to support that. The workflow we support is to run semantic-release to make the release from the release branch when a commit hit the branch.
It's not possible to publish your artifacts outside of semantic-release. If you want to do so you should look for another tool with a more limited scope. 

> Even if the next build would create a release with the two commits, our build fails because we end up trying to override our artifacts pushed with the same version by the previous build.

That's related to the workflow you are using where you do a dry-run but still publish artifact. It's not a semantic-release issue. Again I don't know why you are doing that but that's not the way semantic-release was meant to be used so we won't make changes to accommodate that.

> So I think it is a valid use case to have an option we can configure to allow semantic-release to push tags from a local branch behind the remote that would be disabled by default to keep the backward compatibility.

I don't know what you refer to by backward compatibility. Doing so never really worked. It was allowed by mistake in certain limited cases.

**Comment by rylanc** [comment]
@pvdlg I've got a fairly simple use case where this is failing. I've added semantic-release to a large repo that has a lot of engineers contributing to it. Our CI is set to run only one build at a time and takes ~10 minutes to run. Since the repo is being committed to very often, the semantic-release step is very rarely up-to-date with the remote repo (there's almost always a commit between the time the build is first kicked off and when the release step runs). Reading through the code, things don't look hopeful, but is there any way to make semantic-release work with such an active repo?

**Comment by EvanCarroll** [comment]
Is there a method to error when you get _"The local branch master is behind the remote one, therefore a new version won't be published."_ The rest of my pipeline needs to fail in this case.

**Comment by francescov1** [comment]
@EvanCarroll  I just found https://www.npmjs.com/package/commit-analyzer-fail-on-no-release for this issue

**Comment by WesleyBlancoYuan** [comment]
If someone wonders how to **bypass** the check, and create a release out of a non-HEAD commit, even **this is against the original purpose of semantic-release**:

 - creating a local `release` branch from the commit you want to release: `git checkout -b release <commit-id>`
 - push to remote(required by `semantic-release`; thus the repo cannot be protected): `git push origin release`
 - `npx semantic-release --branches release --no-ci` (`--no-ci` is to bypass the CI environment check, as I am running in local; for CI you don't need it)
 - All the steps which would fail after the HEAD check, will pass now, so for example, creating a new release/summarize the release notes/announce release in Slack channel, will be done.
 - remember to delete the local and remote `release` branch: `git checkout main ; git push origin --delete release; git branch -d release` 

In my case I want to run the following steps after the deployment in our CI, like publish release.
